### PR TITLE
[Doc] [Jobs] Define driver and unify job definitions

### PR DESCRIPTION
**Pull request by architkulkarni**
Signed-off-by: Archit Kulkarni <architkulkarni@users.noreply.github.com>

## Why are these changes needed?
This is a minimal step towards unifying the definition of "Job" in the docs.

Right now there are two types of "Jobs": "Job Submission" jobs, and ordinary Ray drivers (which may or may not use Ray client.)

This PR updates the doc to emphasize that both of these are considered Jobs.  To do this, we need to add a definition for driver, which this PR introduces.

Once we agree on the overall definitions, I can follow up with a more thorough PR that goes through all of the docs and checks for consistency.  (For example, do we want to capitalize Job everywhere?)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (architkulkarni) [inline doc/source/cluster/running-applications/job-submission/index.md:11]: This is awkward.  Open to suggestions

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [inline doc/source/cluster/key-concepts.rst:75]: What are the names of the two types of jobs. One is `Ray Job Submission` and what's the other one?

File: doc/source/cluster/key-concepts.rst
Comment (rkooo567) [inline doc/source/cluster/key-concepts.rst:78]: I am a little confused about this part. I thought we decided to call both driver & submission job as "jobs"? cc @scottsun94

File: doc/source/cluster/key-concepts.rst
Comment (rkooo567) [inline doc/source/cluster/key-concepts.rst:80]: Can you add a link to the Job Submission page?

File: doc/source/cluster/key-concepts.rst
Comment (rkooo567) [inline doc/source/cluster/key-concepts.rst:66]: Why is this not a part of "running_applications" section? I feel like the right structure is

1. Key concept explains this
2. Under running applications, we should demonstrate (with example) 3 ways to run ray workloads to the ray cluster. 1. driver. 2. job submission. 3. ray client?

File: doc/source/cluster/key-concepts.rst
Comment (architkulkarni) [reply doc/source/cluster/key-concepts.rst:78]: That's what was intended here.  Or are you talking about the fact that it says "Ray Job" instead of "job"?

Comment (architkulkarni) [reply doc/source/cluster/key-concepts.rst:78]: Submission jobs also contain a ray driver.  Technically they don't have to (the entrypoint could just be `echo hello` instead of `python my_ray_script.py`), but it seems like an edge case and I'd rather not confuse the reader here

File: doc/source/cluster/key-concepts.rst
Comment (stephanie-wang) [inline doc/source/cluster/key-concepts.rst:70]: I think it would be better to define Ray driver in the same subsection as "Ray job", since we are trying to elevate jobs not drivers.

Also, it is not exactly accurate that the driver needs to call `ray.init()`, since this is now called automatically for certain Ray APIs, like f.remote().

File: doc/source/cluster/key-concepts.rst
Comment (stephanie-wang) [inline doc/source/cluster/key-concepts.rst:78]: This definition is not very precise? It makes it sound like there are user applications that do not include Ray drivers.

Suggested (adapted from whitepaper):
A Ray job is a single application: it is the collection of Ray tasks, objects, and actors that originate from the same entrypoint. The worker that runs the Python entrypoint is known as the *driver* of the job.

File: doc/source/cluster/key-concepts.rst
Comment (stephanie-wang) [inline doc/source/cluster/key-concepts.rst:81]: I probably wouldn't capitalize Submission (lowercasing "jobs" also seems better but we can leave it for later).

```suggestion
Ray Job submission enables users to submit locally developed and tested applications to a
```

File: doc/source/cluster/key-concepts.rst
Comment (stephanie-wang) [inline doc/source/cluster/key-concepts.rst:90]:
```suggestion
For details on these workflows, refer to the :ref:`Ray Jobs guide <jobs-overview>`.
```

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (scottsun94) [reply doc/source/cluster/running-applications/job-submission/index.md:11]: Should we change the structure of the nav?

**Nav:**
Applications guide
- Ray job
   - Get started with ray jobs
   - Key concepts
   - Submit ray jobs to clusters
   - Run ray jobs remotely via ray client
   - Job API reference

For this job page, shall we just provide a more complete definition of ray jobs and add links to those sub-pages above?

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [reply doc/source/cluster/key-concepts.rst:70]: The glossary draft definitions is not correct. Will fix it.

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [reply doc/source/cluster/key-concepts.rst:78]: I feel @scottsun94's understanding is more clear:

Submission is not a type of job, it's just one way to run the job. so we only have one Job concept but multiple ways to run it.

Comment (rkooo567) [reply doc/source/cluster/key-concepts.rst:78]: I think I am happy with Stephanie's definition. Btw isn't this 3 ways, not 2 ways? (job submission, driver, and client )

Comment (rkooo567) [reply doc/source/cluster/key-concepts.rst:78]: Maybe we can call it script?

File: doc/source/cluster/key-concepts.rst
Comment (rkooo567) [inline doc/source/cluster/key-concepts.rst:80]: Actually why don't we structure this way?

```
- [Recommended] Job submission <link>: Blah blah
- Driver: Blah blah
- Client: Blah blah
```

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (rkooo567) [reply doc/source/cluster/running-applications/job-submission/index.md:11]: Actually, I'd like to suggest the alternative; What about combining these two into one section

```
Submit ray jobs to clusters
Run ray jobs remotely via ray client

-> Running ray jobs to the clusters
```

And within one page we have 3 sections, driver + job submission + client. And explain when each of them is useful

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [reply doc/source/cluster/key-concepts.rst:78]: yea, can be 3 ways including ray client.

Comment (scottsun94) [reply doc/source/cluster/key-concepts.rst:78]: 3 ways are fine too. I counted Driver and Client as the interactive way of running ray jobs above.

Comment (scottsun94) [reply doc/source/cluster/key-concepts.rst:78]: script is clearer to me.

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [reply doc/source/cluster/key-concepts.rst:75]: Yea, then do we need a noun called `Ray Job Submission` if it's just a way to run a job.

Or should it be `Submission Job` instead of `Job Submission` to make it clear that it's still a job.

Comment (scottsun94) [reply doc/source/cluster/key-concepts.rst:75]: Ray job submission -> the submission of the job. It depicts this behavior instead of the job itself.
To sum up,
- There is only 1 job concept: Ray job.
- There are 3 ways (run driver on any node, ray client, submit jobs via job api) to run a Ray job
- There is one api called Job API for submitting jobs to remote clusters, listing jobs, retrieving logs, etc.

Is my understanding aligned with yours? @rkooo567 @architkulkarni @stephanie-wang @jjyao

Comment (rkooo567) [reply doc/source/cluster/key-concepts.rst:75]: yeah this sounds good.

Comment (architkulkarni) [reply doc/source/cluster/key-concepts.rst:75]: This sounds good to me.  I'll update this PR accordingly

Comment (jjyao) [reply doc/source/cluster/key-concepts.rst:75]: Yea, that aligns with my understanding as well.

File: doc/source/cluster/key-concepts.rst
Comment (architkulkarni) [reply doc/source/cluster/key-concepts.rst]: Thanks, I've added the new definition

File: doc/source/cluster/key-concepts.rst
Comment (architkulkarni) [reply doc/source/cluster/key-concepts.rst]: Addressed, defined driver inline

**Review by DmitriGekhtman — APPROVED** [review]
code owner stamp, 
currently engaged reviewers are sufficient

**Comment by architkulkarni** [comment]
Thanks for the feedback, it looks like an agreement was reached and I've implemented the changes.

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/index.md]: I've updated this, what do you think of the current update?  The overview page includes two sections, one for Ray Jobs API (recommended) and another for running jobs interactively (driver, and client)

**Review by rkooo567 — APPROVED** [review]

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (rkooo567) [inline doc/source/cluster/running-applications/job-submission/index.md:34]: Is there any doc that explains how to do this? Maybe we should have a link

Comment (rkooo567) [reply doc/source/cluster/running-applications/job-submission/index.md:34]: Ping on this.

Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/index.md:34]: I didn't find any such doc.  I'll link to `ray attach` since it seems like the closest thing.

Comment (scottsun94) [reply doc/source/cluster/running-applications/job-submission/index.md]: `ray attach` won't work for KubeRay right?
Could we actually write a section about it? For people who want to build a platform/service on top of Ray, this is a way people should consider instead of ray client right?

**Comment by scottsun94** [comment]
> One issue with the unified approach we've decided on is that not all "jobs" can be interacted with via the Ray Jobs API. For example, to view the logs of a job or stop a job using the Jobs API, that job must have been started via the Jobs API. I think this is probably still intuitive enough and less confusing than what we had before.

Good point. This is what I wondered about before. 
Is there a path forward to make ray jobs api work with all "jobs"?

@rkooo567 Do other driver jobs work with state api?

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (architkulkarni) [reply doc/source/cluster/running-applications/job-submission/index.md]: @DmitriGekhtman do you know if `ray attach` works for Kuberay, or what the recommended approach is there? (Is it some form of `kubectl exec`)?

@scottsun94 I think users who are running a cluster probably already have a way to run commands on their cluster, almost as a prerequisite (in this case the command we're telling them to run is `python my_ray_script.py`).  So my impression is we don't need a whole section for this, and can just mention SSH or `kubectl` in passing. What do you think?

For people building on top of Ray, we would recommend using the Jobs API or directly running driver scripts.

**Comment by architkulkarni** [comment]
> Good point. This is what I wondered about before. Is there a path forward to make ray jobs api work with all "jobs"?

I don't think so unfortunately, at least not in the short term.  The Jobs API relies on some internal infrastructure that wraps around driver jobs, so the Jobs API won't work for manually run drivers, which are user commands we don't have much control over.  

If users are just running a driver manually, to get some of the benefits of the Ray Jobs API (handling logs, stopping jobs, etc), they can use shell commands that they're already familiar with (e.g. ctrl-C to cancel their script, or use >> to redirect stdout of their script to the file of their choice).

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (scottsun94) [reply doc/source/cluster/running-applications/job-submission/index.md]: Yeah. Probably a short sentence that covers both VM and KubeRay should be fine for now.

**Comment by scottsun94** [comment]
> > Good point. This is what I wondered about before. Is there a path forward to make ray jobs api work with all "jobs"?
> 
> I don't think so unfortunately, at least not in the short term. The Jobs API relies on some internal infrastructure that wraps around driver jobs, so the Jobs API won't work for manually run drivers, which are user commands we don't have much control over.
> 
> If users are just running a driver manually, to get some of the benefits of the Ray Jobs API (handling logs, stopping jobs, etc), they can use shell commands that they're already familiar with (e.g. ctrl-C to cancel their script, or use >> to redirect stdout of their script to the file of their choice).
> 
> (In the Ray Jobs API, `ray job list` does work for both driver jobs and submission jobs currently, and they're tagged with type:Driver or type:Submission)

It makes sense. We could keep it as is and wait for more feedback. Shall we add a note in the jobs api page about this limitation (basically, some of the apis/CLI commands will only work for jobs submitted via the job api)?

**Comment by rkooo567** [comment]
I think it should be okay... better than having more concepts here. I think we can say some APIs are only supported by jobs submitted by ray job submit. Things like stopping jobs & seeing logs is something we can totally implement (if users need it)

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [inline doc/source/cluster/key-concepts.rst:76]: I think we should move `, for interactive development.` to bullet point 2 since we want to de-emphasize ray client.

File: doc/source/cluster/key-concepts.rst
Comment (jjyao) [inline doc/source/cluster/key-concepts.rst:66]: Should it be `Ray Jobs` as the title?

File: doc/source/cluster/running-applications/job-submission/index.md
Comment (jjyao) [inline doc/source/cluster/running-applications/job-submission/index.md:34]: Probably add a `(Recommended)` to the first approach (i.e ssh and run directly)

**Review by jjyao — APPROVED** [review]
Nice!

**Comment by scottsun94** [comment]
Ray diagram

Created a diagram to show the 3 ways of running ray jobs.

**Comment by rkooo567** [comment]
Not a blocker for this PR, but I think we should also have brief description & pros and cons of each job submission model? For example, driver is the best for local dev, users should have ssh access to the head node. Job submission is the best when you submit a script outside the cluster and track the progress. Client is the best when you have jupyter outside the cluster. Something like that

**Comment by rkooo567** [comment]
Also @scottsun94 can you add the second one (run script directly) is called "driver"? We use this term all the time, so it's great to mention that term here.

**Comment by scottsun94** [comment]
> Also @scottsun94 can you add the second one (run script directly) is called "driver"? We use this term all the time, so it's great to mention that term here. 

Sorry. I'm not sure if I fully understand it. Where do you want to add "driver"?

**Comment by rkooo567** [comment]
@architkulkarni this is the link to the picture https://docs.google.com/drawings/d/16IwhnyBw1sUnkQ6U5K4qWIMWVGAntBCLQh-36PmwAH8/edit

**Comment by architkulkarni** [comment]
Sorry `data` codeowners for the accidental review request, this PR doesn't need your review.

@stephanie-wang I've implemented the changes, can you approve this on behalf of ray-docs?

**Comment by architkulkarni** [comment]
Windows failure unrelated (only doc changes)

**Pull request merged**
